### DISCUSSION TITLE
Adding Custom Constants to AcConstants

### DISCUSSION THREAD

**Discussion by ebenezerarunkumar** [nasa/fprime#1555] [Q&A] [answered]
Hi,
I have written a custom component and want to specify the array size of a specific port in the same format as this port in the Active Rate Group: `output port RateGroupMemberOut: [ActiveRateGroupOutputPorts] Sched`. 

I have copied the fprime config folder into my deployment folder, as recommended in a comment ono #1525, and defined my constant, CCSDSOutputPorts, in the `AcConstants.fpp` and `AcConstants.ini` file. I also ran the `update-locs` script in the config folder as well as in the deployment folder.

However, when I build my deployment, I get the following error:

`output port ccsdsOut: [CCSDSOutputPorts] SSDL.CCSDS
                           ^
error: undefined symbol CCSDSOutputPorts
make[3]: *** [SSDL/CCSDSDeframer/CCSDSDeframerComponentAi.xml] Error 1
make[2]: *** [SSDL/CCSDSDeframer/CMakeFiles/SSDL_CCSDSDeframer.dir/all] Error 2
make[1]: *** [CMakeFiles/SSDL_support_CCSDSFramingTest.dir/rule] Error 2
make: *** [SSDL_support_CCSDSFramingTest] Error 2`

I believe the error is because in the `locs.fpp` file within my build directory,`build-fprime-automatic-native`, the constant is not defined, even though it is included in the `locs.fpp` file in my deployment directory. 

Does anyone know why the two `locs.fpp` files are different?

**Comment by ebenezerarunkumar** [comment]
Has anyone else been able to add custom constants into the AcConstants.fpp file and fully compile? Am I missing a step somewhere?

**Comment by ebenezerarunkumar** [comment, marked as answer]
After a little more digging, and with help from, @legounicycler, I believe the issue is that the custom constant I added into the `AcConstants.fpp` and `AcConstants.ini `file is not being referenced and/or parsed. 

In my `settings.ini` file, I have the following lines:

```
; Autocoder constants specific to this deployment
ac_constants: ./config/AcConstants.ini

; This directory contains all of the config hpp files if they
; are not located in the F Prime framework.  It might be config
; or Cfg locally, the framework uses fprime/config or fprime/Fw/Cfg
config_directory: ./config
```
However, during the build step, the output tells me that the Config Directory I set in `config_directory` is not being referenced, even though the correct `AcConstants.ini` is. 
```
-- Autocoder constants file: /Users/earunkumar/fprime-devel/visors-software/SSDL/support/CCSDSFramingTest/config/AcConstants.ini
-- Configuration header directory: /Users/earunkumar/fprime-devel/visors-software/fprime/config
```
It seems as though the cmake variable, `FPRIME_CONFIG_DIR` is not being set correctly based on the settings.ini file, and is always being set to the fprime/config folder. When trying to manually set this var in my deployment CMakeLists.txt, I get the following error:

```
[ERROR] Expected CMake variable FPRIME_CONFIG_DIR to be set to '/Users/earunkumar/fprime-devel/visors-software/fprime/config', was actually set to '/Users/earunkumar/fprime-devel/visors-software/SSDL/support/CCSDSFramingTest/config'. This is usually caused by updating the settings.ini file without purging and regenerating the accompanying build cache.
```
**The only way I was able to get it to work** was to include a CMakeLists.txt file in my config folder, set an fpp file with only my constant in it(not the rest of the constants found in the AcConstants.fpp) as a source file, and include the config folder as a fprime subdirectory.

Is this expected behavior and correct solution to this issue of adding custom constants?

**Reply by LeStarch** [reply]
I think what you have documented here is correct.  I'll take it as an action to fix.  In short:

1. `AcConstants.ini` and `AcConstants.fpp` should be redundant.  Why both?
2. Check the config dir setting and remove the need to specify a separate `AcConstant.ini` setting
3. Make the `constants.fpp` compile and build so that these variables exist for the C++ code

Am I missing anythng?

**Reply by ebenezerarunkumar** [reply]
@LeStarch yes I think the main thing is to have a standardized way to edit the constants used by fprime components, and a way for users to add custom constants

**Reply by ebenezerarunkumar** [reply]
i can put up an issue if that would be helpful

**Reply by LeStarch** [reply]
Yes, a ticket would be great!

**Reply by LeStarch** [reply]
FYI https://github.com/fprime-community/fprime-tools/pull/94  fixes the "cannot set configuration directory" issue. Good catch on that one.

**Comment by legounicycler** [comment]
@LeStarch @bocchino These may be questions that you can answer.

1. How are the `locs.fpp` and `update-locs` files used within F Prime, and is there any use including something similar for a software deployment that has F Prime as a submodule? My suspicion is that these files are used by the developers of F Prime and act as a way to speed up the generate step.
2. How are the various `.ini` files used by F Prime. Since ini is a python format, are these files parsed by python and turned into CMake variables somewhere?
3. How is the `AcConstants.fpp` file created and how is it related to the `AcConstants.ini` file. It seems that in F Prime, any constants you have defined in an `.ini` file must be transformed into a `.fpp` file so that the fpp model has awareness that they exist before generating any code. If that is the case, where/when does the `AcConstants.ini` file get turned into the `AcConstants.fpp` file, or if this is not the case, are you supposed to keep these two files in sync?
